RearrangingCars.java

```diff
@@ -0,0 +1,207 @@
+    /** rearrangingSteps(initial, expected) returns the steps required to arrange the cars
+     *  according to the "expected" rearrangement, starting from the "initial" rearrangement. 
+     *  
+     *  Solution:
+     *      1. Check if the current and the expected rearrangements are valid, if not, then throw an exception.
+     *      2. If the arrays are of length 0 or are exactly equal, there is no steps to be taken, hence return "".
+     *      3. Otherwise, call rearrangingStepsHelper() to get the required series of steps. 
+     *         Return the required steps as a String.           
+     * 
+     * @param initial: an array representing initial rearrangement of the cars
+     * @param expected: an array representing expected rearrangement of the cars
+     * @return String: A string containing a series of steps required to arrange the cars
+     *  according to the "expected" rearrangement, starting from the "initial" rearrangement.
+     * @throws RearrangementException 
+     */
+    public static String rearrangingSteps(int [] initial, int [] expected) throws RearrangementExceptions{
+        
+        StringBuilder str = new StringBuilder();
+        
+        //If the initial or expected rearrangement array is null, then throw an exception.
+        if(initial == null || expected == null){
+            throw new NullRearrangementsException("Either the initial rearrangement or the expected rearrangement is null");
+        }
+        
+        //If the initial and expected rearrangement array do not have same length, then throw an exception.
+        else if(initial.length != expected.length){
+            throw new LengthMismatchException("The length of arrays representing the initial and expected rearrangement are not same");
+        }
+        
+        //If the initial or expected rearrangement array is not a valid permutation of {0,...,array.length-1}, then throw an exception.
+        else if(!isValidPermutation(initial) || !isValidPermutation(expected)){
+            throw new InvalidRearrangementException("The initial or the expected rearrangement is invalid!");
+        }
+        
+        else{
+            //If the initial and expected rearrangement is empty (no slots), then nothing to do, simply return "".
+            if(initial.length == 0){
+                return str.toString();
+            }
+            
+            //If the initial and expected rearrangement are equal, then no steps required, return "".
+            boolean isEqual = true;
+            for(int i=0; i<initial.length; i++){
+                if(initial[i] != expected[i]){
+                    isEqual = false;
+                }
+            }
+            if(isEqual){
+                return str.toString();
+            }
+            else{
+                //Obtain the required series of steps and print them
+                ArrayList<Step> steps = rearrangingStepsHelper(initial, expected);
+                str.append(steps.toString());

```

Comment:
I don't think the StringBulider it's needed. You can return just steps.toString() and "" in cases above.